src/beliefs/discrete.jl

```diff
@@ -1,133 +1,107 @@
-# Maintained by Max Egorov and Zach Sunberg
+# Goals: minimize calls to ordered_states (allocates memory)
 
-"""
-A weight vector to be used as a belief or state distribution.
-
-sum(b)=1 is not enforced at all times, but during the update it is normalized to sum to one.
-
-pdf(b, i) calculates the sum every time it is called. To access the weight directly (for example if you are sure that the sum is 1), use weight(b, i). 
-"""
-mutable struct DiscreteBelief
+# needs pomdp for state_index in pdf(b, s)
+# needs list of ordered_states for rand(b)
+mutable struct DiscreteBelief{P<:POMDP, S}
+    pomdp::P
+    state_list::Vector{S}       # vector of ordered states
     b::Vector{Float64}
 end
+function DiscreteBelief(pomdp)
+    state_list = ordered_states(pomdp)
+    ns = n_states(pomdp)
 
-# Constructor with uniform belief
-DiscreteBelief(n::Int64) = DiscreteBelief(zeros(n) + 1.0/n)
+    b = ones(ns) / ns
 
-mutable struct DiscreteUpdater{P<:POMDP} <: Updater
-    pomdp::P
+    return DiscreteBelief(pomdp, state_list, b)
 end
 
-vec(b::DiscreteBelief) = b.b
 
-Base.length(b::DiscreteBelief) = length(b.b)
-index(b::DiscreteBelief, i::Int64) = i
-weight(b::DiscreteBelief, i::Int64) = b.b[i]
-iterator(b::DiscreteBelief) = filter(i->b[i]>0.0, 1:length(b))
-rand(rng::AbstractRNG, b::DiscreteBelief) = sample(rng, Weights(b.b)) # This will return an integer - seems like it should actually return an object of the state type of the problem
-pdf(b::DiscreteBelief, s::Int) = b.b[s]/sum(b.b) # only works when the state type is integer
+pdf(b::DiscreteBelief, s) = b.b[state_index(b.pomdp, s)]
+
+function rand(rng::AbstractRNG, b::DiscreteBelief)
+    i = sample(rng, Weights(b.b))
+    return b.state_list[i]
+end
 
 function Base.fill!(b::DiscreteBelief, x::Float64)
     fill!(b.b, x)
     return b
 end
 
-function Base.fill!(b::DiscreteBelief, idxs::Vector{Int64}, vals::Vector{Float64})
-    fill!(b.b, 0.0)
-    for i = 1:length(idxs)
-        index = idxs[i]
-        index > 0 ? (b[index] = vals[i]) : nothing 
+# should the belief just store the length so we don't have to count?
```

Comment:
The vector stores its own length, so this is the same as if we stored it (maybe there is an extra pointer dereference, but that is negligible)

Reply:
Awesome, thanks for letting me know. I don't know much when it comes to Julia performance.